src/core/qgsvectorlayercache.cpp

```diff
@@ -231,6 +237,7 @@ void QgsVectorLayerCache::attributeAdded( int field )
 {
   Q_UNUSED( field )
   mCachedAttributes.append( field );
+  mFullCache = false;
```

Comment:
Hmm... `fullCache` is a flag set from externally for providers that do not allow to select by feature id and therefore require a full table scan for every single feature they request. I have a feeling that it's safer to re-scan the table when an attribute is added rather than disabling the full cache or we'll have a bad performance impact repeatedly rather than just once, no?